File: perl/clisoap2.pl
#!/usr/bin/perl
# perl/clisoap2.pl
# Ejecutar: plackup clisoap2.pl -p 5002
# URL: http://localhost:5002/?n=10
# Consume el servicio SOAP y traduce el resultado al español

use strict;
use warnings;
use SOAP::Lite;
use Plack::Request;
use LWP::UserAgent;
use JSON;
use URI;

# Función para traducir usando Google Translate API gratuita
sub google_translate {
    my ($texto, $origen, $destino) = @_;
    $origen //= 'en';
    $destino //= 'es';
    
    my $ua = LWP::UserAgent->new;
    my $uri = URI->new('https://translate.googleapis.com/translate_a/single');
    $uri->query_form(
        client => 'gtx',
        sl     => $origen,
        tl     => $destino,
        dt     => 't',
        q      => $texto
    );
    
    my $response = $ua->get($uri);
    
    if ($response->is_success) {
        my $json = decode_json($response->decoded_content);
        return join('', map { $_->[0] } @{$json->[0]});
    } else {
        return $texto;
    }
}

my $app = sub {
    my $env = shift;
    my $req = Plack::Request->new($env);
    
    my $n = $req->param('n') // 10;
    
    my $wsdl = 'https://www.dataaccess.com/webservicesserver/NumberConversion.wso?WSDL';
    my $cliente = SOAP::Lite->service($wsdl);
    my $resultado_en = $cliente->NumberToWords($n);
    
    my $resultado_es = google_translate($resultado_en, 'en', 'es');
    
    return [
        200,
        ['Content-Type' => 'text/plain; charset=UTF-8'],
        [ $resultado_es ]
    ];
};

# IMPORTANTE: Devolver la referencia de código
$app;

File: nodejs/clisoap2.js
// nodejs/clisoap2.js
// Uso: node clisoap2.js  -> http://localhost:3002/?n=10
// Consume el servicio SOAP y traduce el resultado al español

const http = require('http');
const url = require('url');
const soap = require('soap');
const { translate } = require('@vitalets/google-translate-api');

const WSDL = 'https://www.dataaccess.com/webservicesserver/NumberConversion.wso?WSDL';

const server = http.createServer(async (req, res) => {
  const { query } = url.parse(req.url, true);
  const n = parseInt(query.n) || 0;

  try {
    // 1. Llamada SOAP
    const cliente = await soap.createClientAsync(WSDL);
    const [respuesta] = await cliente.NumberToWordsAsync({ ubiNum: n });
    const en = respuesta.NumberToWordsResult;

    // 2. Traducción inglés -> español
    const result = await translate(en, { from: 'en', to: 'es' });
    const es = result.text;

    res.writeHead(200, { 'Content-Type': 'text/plain; charset=utf-8' });
    res.end(es);
  } catch (err) {
    console.error('Error:', err);
    res.writeHead(500);
    res.end(`Error: ${err.message}`);
  }
});

server.listen(3002, () => console.log('http://localhost:3002/?n=10'));